### ISSUE TITLE
Possible Issue Mixing Spatial Resolutions

### ISSUE THREAD

**Issue by GOESguy** [ssec/sift#314]
I have created a new ABI "Rocket Plume" RGB that uses 2km (R and G) and 0.5km (B component). I have run the recipes in both SIFT and geo2grid. I get the faint blue pixel (the plume) showing up near the John C. Stennis Space Center in Mississippi in the geo2grid, but not the SIFT. See the attached image (geo2grid on the left and SIFT on the right). The GOES-16 CONUS day/time and recipe are part of the image, so this can be re-produced. If not, let me know. Jordan wonders if somehow the larger 2km pixels are over-plotting the finer resolution ABI band 2 visible? Thanks. 
![where_is_blue_in_sift_18March2021_CONUS_2040UTC_Stennis](https://user-images.githubusercontent.com/87716578/126363046-d178e8f9-1d3f-46d6-9ce0-ccfd2cfd64d1.png)

**Comment by djhoese** [comment]
You are pretty zoomed out in the SIFT window. I'm guessing it still doesn't show up even if you zoom in further? If it is a single pixel then it will require loading the full resolution data in SIFT which doesn't happen until it is zoomed in enough. I'm sure you tried, but just making sure.

Just for full comparison, could you paste your composite and enhancement YAML from Geo2Grid here? Maybe there is something small happening in Satpy that is able to bring out this pixel that SIFT isn't able to do.

**Comment by ScottLindstrom** [comment]
Is that blue spot coming from the isolated bright spot in the vis?  See attached -- zoomed.

I've tweaked the gammas -- small values.  The geen has 1, 1, 0.3 for gammas, the yellow has all 0.3; I can't get the blue region in the geo2grid to show up in the same place.
________________________________

**Comment by djhoese** [comment]
@ScottLindstrom I don't think github (where your email reply ends up) accepts attachments. We don't see an attachment here (on github) or in the email notification about this comment.

**Comment by ScottLindstrom** [comment]
![TimSPlume_Gamma1 1 0 3](https://user-images.githubusercontent.com/16156223/126375251-a412ac5d-e144-4fa1-86f7-3c069e1c29e5.png)
![TimSPlume_Gamma0 3](https://user-images.githubusercontent.com/16156223/126375252-761bbefd-28fd-491a-8444-eddcd6cc8e7e.png)

(Sure.  Make me reset my password)

**Comment by GOESguy** [comment]
Yes, the blue dot is the plume from the ABI band 2 (vis).

Attached:

  *   image I tried to attached
  *   composite abi.yaml
  *   enhancements abi.yam

Thanks.


--
Timothy J. Schmit
NOAA/NESDIS Center for SaTellite Applications and Research (STAR)
Advanced Satellite Products Branch  (ASPB)

1225 West Dayton Street   Madison, WI 53706

["The Birthplace of Satellite Meteorology"]

***@***.******@***.***>    or   ***@***.******@***.***>

(608) 263-0291 (voice)
(608) 262-5974 (facsimile)

________________________________

**Comment by GOESguy** [comment]
In this case I'm pretty zoomed in.
Attached are the 3 bands and the RGB.
Sorry about the blue cursor, but the plume is to it's NorthEast.
Thanks.


--
Timothy J. Schmit
NOAA/NESDIS Center for SaTellite Applications and Research (STAR)
Advanced Satellite Products Branch  (ASPB)

1225 West Dayton Street   Madison, WI 53706

["The Birthplace of Satellite Meteorology"]

***@***.******@***.***>    or   ***@***.******@***.***>

(608) 263-0291 (voice)
(608) 262-5974 (facsimile)

________________________________

**Comment by djhoese** [comment]
Attachments do **NOT** work through github comment email replies.

**Comment by djhoese** [comment]
Ok I had time to investigate this and the answer for why this is happening is disappointingly simple: SIFT will use the coarsest resolution to make its decisions on RGBs. Looking through the code this is only an issue for the RGBs and **not** algebraic layers. I did not leave myself great documentation for why I did this, but if I had to guess this was done in the rush to get SIFT working with RGBs originally and doing it this way was much easier.

There are various design reasons for why this was done and there are multiple ways to do it the finer resolution way. There are pros/cons to each approach which I'll lay out here (FYI @rayg-ssec):

1. Current situation: Uses less memory (GPU and CPU) as we don't have to replicate coarse resolution pixels to match the shape of the fine resolution pixels (ex. replicate 2km data by 4x in both dimensions). The code is simpler and allows us to do simple "striding" with fine resolutions to get them to the coarser resolutions (ex. `ch02_data[::2, ::2]` is now at 1km resolution). Downside is that the RGB image will always use a lower resolution version of the data when mixing resolutions.
2. Replicate (my preferred): Code stays very similar but with small extra complexity to replicate pixels instead of striding them. This biggest downside to this option is that any RGBs that combine 2km with 1km/500m, or 1km with 500m data will result in 4x or 2x the memory usage.
3. Individual texture coordinates: This basically means telling the GPU the coordinates/locations for each R, G, and B piece of image data. This means a lot more complexity in the code and a lot less reused code (both in SIFT's own code and the code it uses from vispy).

In summary: The behavior described in this bug report are expected, I just forgot that they were.

**Comment by GOESguy** [comment]
Dave,

Thanks for the investigation. (Jordan was right, again.)

What does geo2grid do? Your option #2? If so, might be another reason for option #2.

Thanks,
Tim


--
Timothy J. Schmit
NOAA/NESDIS Center for SaTellite Applications and Research (STAR)
Advanced Satellite Products Branch  (ASPB)

1225 West Dayton Street   Madison, WI 53706

["The Birthplace of Satellite Meteorology"]

***@***.******@***.***>    or   ***@***.******@***.***>

(608) 263-0291 (voice)
(608) 262-5974 (facsimile)

________________________________

**Comment by djhoese** [comment]
Geo2Grid has a couple options. If you use the `native` resampler (default) and resample to the "MAX" grid (also default) then yes it is doing something very similar to option 2 I described above (replicate coarser bands to be the same shape as the finer bands). If you do `-g MIN` then it will aggregate/average the finer (500m) pixels to come up with the coarser version of the data. This is different than SIFT which simply strides (give me every N pixels).

**Comment by djhoese** [comment]
@jgerth Oh that's a good idea! It of course is slightly more complicated to bounce between the two cases (I think), but something as simple as "if we would stride the coarsest band by 1 then switch to replicating". This would be nice because theoretically I think you would have 2 extra zoom levels then based on the finer resolution data (2km versus 500m). For example, if we are striding the 2km data by 1 then each pixel represents 2km by 2km. The 500m data would be strided by 4 in this case so a stride by 2 and by 1 should still be possible.

I'll have to think about this. I should also talk to the EUMETSAT folks about this.
